hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/upgrade/HDDSLayoutFeatureCatalog.java

```diff
@@ -39,7 +40,11 @@
 
     private int layoutVersion;
     private String description;
-    private Optional< ? extends HDDSUpgradeAction> hddsUpgradeAction =
+
+    private Optional<? extends HDDSUpgradeAction> scmUpgradeAction =
+        Optional.empty();
+
+    private Optional<? extends HDDSUpgradeAction> datanodeUpgradeAction =
```

Comment:
I am wondering whether the organization of modules will cause a problem here. The DatanodeStateMachine class is in container-service module, while this class (HDDSLayoutFeatureCatalog) is in hdds-common. The common module is usually shared across the the more specific modules. Where does one create an upgrade action for the Datanode? If the action is in container-service, then it cannot be accessed in hdds-common, if the action is in common, then it cannot access the DatanodeStateMachine class.

Reply:
Yes. I have updated the PR to add datanode action and scm action. There is an integration test for this, thats working but thats in a separate PR for integration test.

Reply:
New approach looks good to me. Can we fix the CI issues? Also, do we need to include the sample SCM and DN action that you have added in the latest commit? Maybe we could use a different implementation that accepts an empty list (enum does not).

Reply:
Yes, uploaded the new changes to address CI failures. We can leave sample SCM and DN actions to be used as an example in future upgrades. This also makes it a self validating release.

Reply:
@prashantpogde I am ok with that. If we are going to add these No-Op actions for the first upgrade release, I would suggest renaming the action to something more intuitive rather than "DatanodeAction1" and "SCMAction1". I am OK with doing these changes in a follow up JIRA.